src/allotropy/allotrope/schema_parser/schema_cleaner.py

```diff
@@ -0,0 +1,621 @@
+                    for add_schema in schemas_list
+                ]
+                if new_schemas
+                else [[schema] for schema in schemas_list]
+            )
+        return {key: [{"allOf": schemas} for schemas in new_schemas]}
+
+    def _combine_allof_schemas(
+        self, schemas: list[dict[str, Any]]
+    ) -> Union[dict[str, Any], list[dict[str, Any]]]:
+        if not all(_is_class_schema(schema) for schema in schemas):
+            if any(_is_class_schema(schema) for schema in schemas):
+                msg = f"_combine_allof_schemas can only be called with a list of object schema dictionaries: {schemas}"
+                raise AssertionError(msg)
+
+            return schemas
+
+        return self._try_combine_schemas(schemas)
+
+    def _combine_allof(self, values: list[Any]) -> dict[str, Any]:
+        if not all(isinstance(value, dict) for value in values):
+            msg = "Unhandled case: expected every item in an allOf to be a dictionary"
+            raise AssertionError(msg)
+
+        if len(values) == 1 or _all_values_equal(values):
+            return validate_dictionary_type(values[0])
+
+        # If we are trying to combine tQuantityValue with tQuantityValue units, we can do so if there are
+        # not conflciting units.
+        if all(
+            _is_ref_schema(schema) and "QuantityValue" in schema["$ref"]
+            for schema in values
+        ):
+            unique = {_get_def_name(schema["$ref"]) for schema in values} - {
+                "tQuantityValue"
+            }
+            if len(unique) == 1:
+                return {"$ref": f"#/$defs/{next(iter(unique))}"}
+            msg = f"Unable to combine multiple different tQuantityValue references: {values}"
+            raise AssertionError(msg)
+
+        # datamodel-codegen can not handle oneOf nested inside allOf. Fix this by reversing the order,
+        # making a oneOf with each possible product of allOf
+        if any("oneOf" in value for value in values):
+            return self._clean_schema(self._invert_allof(values, "oneOf"))
+
+        # This must come after fixing oneOf because oneOf may break into multiple quantity values.
+        if self._is_quantity_value(values):
+            return self._fix_quantity_value_reference(values)
+
+        # Deference values and check for oneOf/anyOf inversion again.
+        derefed_values = self._dereference_values(values)
+        if any("oneOf" in value for value in derefed_values):
+            return self._clean_schema(self._invert_allof(derefed_values, "oneOf"))
+        if any("anyOf" in value for value in derefed_values):
+            return self._clean_schema(self._invert_allof(derefed_values, "anyOf"))
+
+        # If any object in the allOf has required fields, we must combine them in order for the generation
+        # script to generate valid dataclasses. This is because dataclass inheritance with optional fields
+        # is broken in python < 3.10.
```

Comment:
So, there are two answers to this:

1. For compatibility with the monolith (which when I _started_ this, was python 3.9 😭)
2. Even after the upgrade, I'm not sure this will work without further adjustment, because you still need special qualification in the inheritance. I do plan to see if it the generation script will handle this after we upgrade python, which we will do someday soon.